### PR TITLE
Galasactl launches local tests in local JVMs

### PR DESCRIPTION
**Pull request by techcobweb**
## Objectives
This PR adds the ability to call the galasactl command-line tool using syntax
`galasactl runs submit local ...` 

## Tracking issue
https://github.com/galasa-dev/projectmanagement/issues/1322

## Tasks
- [x] throttle file not used if throttlefilename not set
- [x] Launcher interface added to separate JVM launcher from remote launcher
- [x] dependencies include openapi-generator-cli which is used by build-locally.
- [x] build-locally --clean forces re-generation of code from yaml file and dependency download.
- [x] pushed submit runs functionality into runs package, so it can be re-used by a local variant.
- [x] makefile cleaned up a bit
- [x] when tests fail, say how many failed in the error message.
- [x] explain what the syntax is for a class parameter better
- [x] hide the apiclient within the remote implementation of a launcher completely.
- [x] separate out all bootstrap functions into a bootstrap module with unit tests added
- [x] galasa errors don't produce a stack trace. Unexpected non-galasa errors do create one in the log file.
- [x] new jvmlauncher implements the launcher interface
- [x] type holding command parameters moved into utils so it can be used from everywhere
- [x] added launcher interface, to (hopefully) treat local and remote launches the same.
- [x] env module mocks out the environment variables.
- [x] runID and ras folder location extracted from JVM trace as a test runs.
- [x] shutdown of framework extracted from JVM trace as test runs.
- [x] test ending for local test causes sleep to be broken and new test scheduled immediately.
- [x] test results read from file in RAS file store and fed back to submitter for reporting.
- [x] Galasa errors with proper error text
- [x] Unit tests for JVM launcher

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by techcobweb** [comment]
@jadecarino Thanks for volunteering to review. I've scheduled a meeting for Thursday, but if I escape the GHE education early on Weds I'll give you a shout.
Ideally I would not have made such a big change, but it basically required re-factoring most of the submit code, and removing all the panic() calls so that it can be unit tested.

File: pkg/launcher/jvmLauncher.go [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+		testName := localTest.testRun.GetName()
+		if localTest.isCompleted() {
+			log.Printf("GetRunsByGroup: localTest %s is complete.\n", testName)
+		} else {
+
+			// Update the test status by reading the json file if we can.
+			localTest.updateTestStatusFromRasFile()
+
+			if localTest.isCompleted() {
```
Comment (jadecarino) [inline pkg/launcher/jvmLauncher.go:222]: Should this be !localTest.isCompleted()

File: docs/generated/galasactl_runs_submit_local.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+### Options inherited from parent commands
+
+```
+  -b, --bootstrap string           Bootstrap URL
+  -g, --group string               the group name to assign the test runs to, if not provided, a psuedo unique id will be generated
+  -l, --log string                 File to which log information will be sent. Any folder referred to must exist. An existing file will be overwritten. Specify "-" to log to stderr. Defaults to not logging.
+      --noexitcodeontestfailures   set to true if you don't want an exit code to be returned from galasactl if a test fails
+      --override strings           overrides to be sent with the tests (overrides in the portfolio will take precedence). Each override is of the form 'name=value'. Multiple instances of this flag can be used. For example --override=prop1=val1 --override=prop2=val2
+      --poll int                   Optional. The interval time in seconds between successive polls of the ecosystem for the status of the test runs. Defaults to 30 seconds. If less than 1, then default value is used. (default 30)
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply docs/generated/galasactl_runs_submit_local.md:37]: yes this needs to change.

File: pkg/cmd/runsPrepare.go [hunk 3]
```diff
@@ -66,12 +72,21 @@ func executeAssemble(cmd *cobra.Command, args []string) {
 		testOverrides[key] = value
 	}
 
-	apiClient, err := api.InitialiseAPI(bootstrap)
+	// Load the bootstrap properties.
+	var urlService *api.RealUrlResolutionService = new(api.RealUrlResolutionService)
+	bootstrapData, err := api.LoadBootstrap(fileSystem, env, bootstrap, urlService)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/cmd/runsPrepare.go:79]: yes. needs fixing.

File: pkg/errors/errorMessage.go [hunk 4]
```diff
@@ -114,6 +114,10 @@ var (
 	GALASA_ERROR_JAVA_HOME_BIN_MISSING          = NewMessageType("GAL1052E: Folder '%s' is missing. JAVA_HOME environment variable should refer to a folder which contains a 'bin' folder.", 1052)
 	GALASA_ERROR_JAVA_PROGRAM_PRESENCE_FAIL     = NewMessageType("GAL1053E: Failed to determine if '%s' exists. Reason is '%s'", 1053)
 	GALASA_ERROR_JAVA_PROGRAM_MISSING           = NewMessageType("GAL1054E: Program '%s' should exist. JAVA_HOME has been set incorrectly.", 1054)
+	GALASA_ERROR_FAILED_TO_READ_BOOTSTRAP_FILE  = NewMessageType("GAL1055E: Failed to read bootstrap file '%s'. Rason is '%s'", 1055)
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/errors/errorMessage.go:117]: Rason needs correcting

File: pkg/launcher/jvmOutputProcessor.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		isShutdownDetected := detectShutdown(stringToSearch)
+		if isShutdownDetected {
+			isAlertable = true
+		}
+
+		if isAlertable {
+			// Now alert anyone who may be listening on the go channel.
+			processor.publishResultChannel <- "ALERT"
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/launcher/jvmOutputProcessor.go:96]: read up on go "channels".

File: pkg/launcher/jvmLauncher.go (hunk 1 above)
Comment (techcobweb) [reply pkg/launcher/jvmLauncher.go:222]: Corrected in following PR on overrides.

**Review by jadecarino — APPROVED** [review]

File: pkg/cmd/runsPrepare.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/cmd/runsPrepare.go:79]: Added it into the api.LoadBootstrap call to make sure it delivered a Galasa error.

File: docs/generated/galasactl_runs_submit_local.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (techcobweb) [reply docs/generated/galasactl_runs_submit_local.md]: Inherited text changed to ... `polls the test runs status`.
Fix in this PR.

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**